### ISSUE TITLE
[feedback] 3D Leitner Boxes

### ISSUE THREAD

**Issue by meolox** [fabd/kanji-koohii#25]
ファブリス,

Would you consider keeping the 3D look for the leitner boxes on the desktop site and have it shrink down to the flat 2D bars for mobile? It's picky I know but I feel the 3D look makes it feel like a physical collection of cards piling up which is a bit more motivating for me than the flat 2D bars currently on the staging site, that seems like a statistic on a bar chart rather than an indication of progress that has depth....sorry flat design purists.

**Comment by fabd** [comment]
Hehe, rest assured I am not particularly fond of flat design. However it makes it much easier to focus with the mobile updtates. Over time I want to add back some details like subtle shadows or subtle gradients.

I am sad to see the isometric bars go too.

It could work with CSS only? I'm not sure. If someone wants to fiddle with that on jsfiddle or codepen, I could add it in. I think CSS border property can be used to make the slanted edges.

I don't want to use CANVAS because that won't work with VueJS approach, afaik.

The last option, compatible with VueJS is to use SVG, The 3D Leitner graph was actually made with SVG and worked all the way down to IE6 with [Raphel library](https://github.com/DmitryBaranovskiy/raphael). 

I'm not too fond about including yet another library on top of YUI2 and VueJS, but it looks like SVG is [more standard nowadays (caniuse.com)](http://caniuse.com/#feat=svg)... so maybe this is a workable solution.

**TLDR** Hard to prioritize this for now so I have to label `nice-to-have` but this is def; something people can help me with, particularly since I shared this codePen some time ago!

http://codepen.io/fabd/pen/Ypjjjg

**If anybody wants to help:**

1. Make it show 3D bars with SVG elements
2. Have an option in the Vue instance that switches to 2D bars for mobile (unless it can fit a smaller display)

**Comment by fabd** [comment]
In case anybody forgets, when I have to update the live site with the mobile / responsive update...

'member the 3D bars =)

![fwb16vwigsgfj2vl9dscdggxr46mkrercrwp8hfu8brti4lxiaaaaasuvork5cyii](https://cloud.githubusercontent.com/assets/169391/21999810/2406bab0-dc3c-11e6-9417-9a79725ae241.png)

Not but I do not mean to jest, this is the reference in case someone wants to play with the CodePen.

**Comment by faneca** [comment]
Hi, I can come up with two CSS-only solutions:

## `box-shadow`
A fast-n-ugly way could be to add a box-shadow like this
```
.box .bar {
    ...
    box-shadow: 1px -1px, 2px -2px, 3px -3px, 4px -4px, 5px -5px, 6px -6px, 7px -7px; /* and so on for deeper bars */
}
```
and then control the color with the `color: ...` property of the elements (either in a class or in each bar element -- I've seen the colors are applied directly to the elements, but would recommend adding to each one just the pertinent class, and let those classes' CSS give the bars its respective colors).
Note I've only tested this on Firefox: according to the [box-shadow specification](https://www.w3.org/TR/css3-background/#box-shadow) it should be correct, but [caniuse](http://caniuse.com/#search=box-shadow) says that maybe you should also add a 3rd and 4th length elements of exactly "1px" each (e.g. `box-shadow: 1px -1px 1px 1px, 2px -2px 1px 1px, 3px -3px 1px 1px, ...`) for better compatibility.
  - Pros: easy, pretty much supported everywhere.
  - Cons: Instead of two shades, we get just one color.

## `::before` and `::after` with a `transform: skew(...)`
If we prefer to keep the two differentiated shades (three if we count the foreground face of the prisms), we could resort to `::before` and `::after` pseudo-elements with a `transform: skew(...)` applied to them. For instance:
```
.box .bar::before {
	content: "";
	position: absolute;
	width: 24px;
	height: 100%;
	left: 100%;
	background-color: white; /* you should change this for the different colors */
	transform: skew(0, -20deg);
	transform-origin: top left;
}
.box .bar::after {
	content: "";
	position: absolute;
	width: 100%;
	height: 9px;
	bottom: 100%;
	background-color: lightblue; /* you should change this for the different colors */
	transform: skew(-70deg);
	transform-origin: bottom right;
}
```
where adding the different vendor prefixes to the `transform...` properties could also be a good idea for better compatibility.
  - Pros: We get the colorful experience ;-)
  - Cons: I usually tend to have problems with the `transform` properties on non-standard OS' (*cough* windows *cough*) default browser (*cough* explorer *cough* edge *cough* *cough*), and having to add the prefixes `-moz-`, `-webkit-` and family is tedious and ugly.

Hope this helps.

**Comment by fabd** [comment]
Very cool faneca!

**Here is solution one** as a CodePen (single color)

http://codepen.io/fabd/pen/YNpGvK

**Comment by fabd** [comment]
**Re: solution 2**

Have to try it (I'm in some 2 day training, checking this as time avail).

FYI : less then 2% Kanji Koohii visitors use Internet Explorer, 89% *of which* use IE10+ (from what I remember).

**TLDR : if it doesn't show in IE, it's OK** as long as it doesn't break. Must support: Android 4.4+, iOS Safari.

**Comment by fabd** [comment]
**Working on the pen here for solution 2**

It looks like this could work. Figuring out how to put the colours via VueJs.

Thought I should use 45 degrees to avoid gaps between the pixels.

http://codepen.io/fabd/pen/PWbbwp/

**Comment by fabd** [comment]
Very cool!! @meolox let me know if this proof of concept works on your desktop! (colours a bit different)

http://codepen.io/fabd/pen/PWbbwp/

**Comment by fabd** [comment]
@faneca @meolox

Thank you faneca!

Recompiled the CodePen with [Bublé](https://buble.surge.sh/) to remove ES6 (let, const, ...) and it's **working in IE11** , though it shows a 1 pixel gap in some places, but I can live with that.

http://codepen.io/fabd/pen/PWbbwp/

When I'm home I'll try to merge it in the site. Still needs some tweaks, to see how I can toggle the isometric elements when the space is too tight. I feel like the isometric display could be used on mobile as well, if I summarize the display in eg. 5 boxes instead of 8.

Then I'll post it on [the Gitter chat](https://gitter.im/kanji-koohii/feedback) for testing on actual devices.

**Comment by fabd** [comment]
Teehee we can use a 2D oldschool depth effect on smaller screens too, i like this (CodePen updated with a @media rule):

![untitled](https://cloud.githubusercontent.com/assets/169391/22024983/35e86c52-dccc-11e6-96fd-ebe757b3ed3f.png)

**Comment by faneca** [comment]
Glad to help! ;-)

**Comment by meolox** [comment]
<img width="141" alt="screen shot 2017-01-17 at 21 18 18" src="https://cloud.githubusercontent.com/assets/25160633/22040362/7f6b6284-dcfa-11e6-8201-51e957786de5.png">

You both beat me to it! I was working on similar to what @faneca suggested, sightly different, drawing 3 boxes (top, side, front) and using skew transform to form a cube but it was truly ugly and didn't like Safari on iOS.

Glad they're back in better code than I could do 👍

**Comment by fabd** [comment]
Forgot to check on my iPad yesterday, could someone confirm the [CodePen](http://codepen.io/fabd/pen/PWbbwp/) works in iOS Safari? (I didn't yet use css vendor prefixes in this codepen).

**Comment by fabd** [comment]
Merged into **staging** for next live update.

Also added depth and "pressed" (:active) state to the SRS buttons.

Tested on my iPad / Safari iOS 8 and Internet Exploder 11.

( If there are issues, add comments, a closed issue can be re-opened )
